checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -363,57 +428,132 @@ private static int getGroupCount(Pattern p) {
+         * Traverse until you find the corresponding '\E', then resume normal traversal.
+         */
+        final int length = regexp.length();
+        for (int i = 0; i < length; i++) {
+            if (regexp.charAt(i) == '(') {
+                if (!escaped) {
+                    if (i != length - 1) {
+                        if (regexp.charAt(i + 1) == '?') {
+                            if (i < length - 2
+                                    && regexp.charAt(i + 2) == '<') { // named capturing group.
+                                group += 1;
+                                if (i != 0 && regexp.charAt(i - 1) == '|') {
+                                    nonNullGroups.remove(Integer.valueOf(group));
+                                }
+                                openingIndices.push(group);
+                                openingWasGroup.push(true);
+                            } else { // non capturing group.
+                                openingWasGroup.push(false);
+                            }
+                        } else { // unnamed capturing group.
+                            group += 1;
+                            if (i != 0 && regexp.charAt(i - 1) == '|') {
+                                nonNullGroups.remove(Integer.valueOf(group));
+                            }
+                            openingIndices.push(group);
+                            openingWasGroup.push(true);
+                        }
+                    }
+                } else {
+                    escaped = false;
                 }
-                group += 1;
-                if (i != 0 && regexp.charAt(i - 1) == '|') {
-                    nonNullGroups.remove((Integer) group);
+            } else if (regexp.charAt(i) == ')') {
+                if (!escaped) { // ending of a construct.
+                    boolean closesGroup = openingWasGroup.pop();
+                    if (closesGroup) {
+                        int value = openingIndices.pop();
+                        if (i != regexp.length() - 1
+                                && "?|*".contains(Character.toString(regexp.charAt(i + 1)))) {
+                            nonNullGroups.remove(Integer.valueOf(value));
+                        } else if (i < length - 2
+                                && regexp.charAt(i + 1) == '{'
+                                && regexp.charAt(i + 2) == '0') {
+                            nonNullGroups.remove(Integer.valueOf(value));
+                        }
+                    }
+                } else {
+                    escaped = false;
                 }
-                openingIndices.push(group);
-            } else if (!escaped && !squareBracketOpen && regexp.charAt(i) == ')') {
-                if (notCapturing) {
-                    notCapturing = false;
-                    continue;
+            } else if (regexp.charAt(i) == '\\') {
+                escaped = !escaped;
```

Comment:
The handling of the `escaped` variable is verbose, and near the handling of 'Q' seems buggy.
I think it would be cleaner to eliminate the `escaped` variable, and right here handle the two relevant cases.
 * If the next character is 'Q', increment `i` to the "E" in the next "\E".
 * Otherwise, increment `i`.

I think that might be all that is needed, but let me know if that is missing something.  The little loop for character class ('[') handling could be similarly handled without an `escaped` variable.